src/Tablebot/Plugin/Netrunner.hs

```diff
@@ -60,11 +59,62 @@ fuzzyQueryCard pairs = closestValueWithCosts editCosts unpackedPairs . unpack
+searchCards :: NrApi -> [(String, Char, String)] -> Maybe [Card]
+searchCards _ [] = Nothing
+searchCards NrApi {cards = cards} pairs = Just $ nubBy cardEq $ foldr filterCards cards $ packValues pairs
   where
-    normalizedText = normalize NFD (toLower txt)
+    packValues :: [(String, Char, String)] -> [(String, Char, Text)]
+    packValues = map (\(k, sep, v) -> (k, sep, pack v))
+    cardEq :: Card -> Card -> Bool
+    cardEq a b = title a == title b
+    filterCards :: (String, Char, Text) -> [Card] -> [Card]
+    filterCards ("x", sep, x) = filterText sep stripped_text x
+    filterCards ("a", sep, x) = filterText sep flavor x
+    filterCards ("e", sep, x) = filterText sep pack_code x
+    -- filterCards ("c", sep, x) =
+    filterCards ("t", sep, x) = filterText sep type_code x
+    filterCards ("f", sep, x) = filterText sep faction_code x
+    filterCards ("s", sep, x) = filterText sep keywords x
+    filterCards ("d", sep, x) = filterText sep side_code x
+    filterCards ("i", sep, x) = filterText sep illustrator x
+    filterCards ("o", sep, x) = filterInt sep cost x
+    filterCards ("g", sep, x) = filterInt sep advancement_cost x
+    filterCards ("m", sep, x) = filterInt sep memory_cost x
+    filterCards ("n", sep, x) = filterInt sep faction_cost x
+    filterCards ("p", sep, x) = filterInt sep strength x
+    filterCards ("v", sep, x) = filterInt sep agenda_points x
+    filterCards ("h", sep, x) = filterInt sep trash_cost x
+    -- filterCards ("r", sep, x) cs =
+    filterCards ("u", sep, x) = filterBool sep uniqueness x
+    -- filterCards ("b", sep, x) cs =
+    -- filterCards ("z", sep, x) cs =
+    filterCards _ = id
+    filterText :: Char -> (Card -> Maybe Text) -> Text -> ([Card] -> [Card])
+    filterText ':' f x = filter ((isInfixOf $ standardise x) . standardise . (fromMaybe "") . f)
+    filterText '!' f x = filter (not . (isInfixOf $ standardise x) . standardise . (fromMaybe "") . f)
+    filterText _ _ _ = id
+    filterInt :: Char -> (Card -> Maybe Int) -> Text -> ([Card] -> [Card])
+    filterInt sep f x = case readMaybe $ unpack x of
+      Nothing -> id
+      Just x' -> filter (\c -> (fromMaybe False) $ ((toEquality sep) x') <$> f c)
+    toEquality :: Char -> (Int -> Int -> Bool)
+    toEquality ':' = (==)
+    toEquality '!' = (/=)
+    toEquality '>' = (<) -- These are right, see the @filterInt@ for their application
+    toEquality '<' = (>)
+    toEquality _ = \_ _ -> True
+    filterBool :: Char -> (Card -> Maybe Bool) -> Text -> ([Card] -> [Card])
+    filterBool ':' f _ = filter (fromMaybe False . f)
+    filterBool '!' f _ = filter (fromMaybe True . f)
+    filterBool _ _ _ = id
+
+-- | @pairsToQuery@ takes a set of search query pairs ands turns it into a link
+-- to an equivalent search on NetrunnerDB.
+pairsToQuery :: [(String, Char, String)] -> Text
+pairsToQuery pairs = "<https://netrunnerdb.com/find/?q=" <> replace " " "+" (intercalate "+" queries) <> ">"
```

Comment:
that's a very cool use of `intercalate`
although this is the source of the bug where a fuzzy search you've applied (e.g. `f:weyland`) isn't applied in the link - so the `more` link gives a bad NRDB page

Reply:
This has now been solved, but highlights a discrepency in that my search is non-exclusive (`f:n` matches any faction name containing an `n` - which is most of them) whereas NetrunnerDB's search is exclusive (`f:n` autocompletes to `f:nbn`).

Should the search algorithm I use be made to mirror this?

Reply:
As discussed, the search algorithm will be changed.